### PR TITLE
wallet: Improve CWallet:MarkDestinationsDirty

### PR DESCRIPTION
**Pull request by promag**
Improve `CWallet:MarkDestinationsDirty` by skipping transactions that already have the cache invalidated. Skipping a transaction avoids at worst case extracting all output destinations.

### CODE REVIEW

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -310,6 +310,7 @@ class CWalletTx
     enum AmountType { DEBIT, CREDIT, IMMATURE_CREDIT, AVAILABLE_CREDIT, AMOUNTTYPE_ENUM_ELEMENTS };
     CAmount GetCachableAmount(AmountType type, const isminefilter& filter, bool recalculate = false) const;
     mutable CachableAmount m_amounts[AMOUNTTYPE_ENUM_ELEMENTS];
+    mutable bool m_is_cache_empty{true};
```
Comment (instagibbs) [inline src/wallet/wallet.h]: a comment for semantics and uses would be nice

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -431,11 +432,13 @@ class CWalletTx
     //! make sure balances are recalculated
     void MarkDirty()
     {
+        fChangeCached = false;
+        if (m_is_cache_empty) return;
```
Comment (achow101) [inline src/wallet/wallet.h]: I don't think we should have an early return. The reset shouldn't be very expensive, and I think it's better to be sure that the cache is cleared when `m_is_cache_empty = true`

Comment (promag) [reply src/wallet/wallet.h]: Done.

File: src/wallet/wallet.h (hunk 1 above)
Comment (promag) [reply src/wallet/wallet.h]: @instagibbs done (I think?)

**Comment by fjahr** [comment]
Code review ACK c9a314c6dd3bf48f5bc58e606470b39a3c963724

File: src/wallet/wallet.h [hunk 3]
```diff
@@ -313,6 +313,13 @@ class CWalletTx
     enum AmountType { DEBIT, CREDIT, IMMATURE_CREDIT, AVAILABLE_CREDIT, AMOUNTTYPE_ENUM_ELEMENTS };
     CAmount GetCachableAmount(AmountType type, const isminefilter& filter, bool recalculate = false) const;
     mutable CachableAmount m_amounts[AMOUNTTYPE_ENUM_ELEMENTS];
+    /**
+     * This flag is true if all m_amounts caches are empty. This is particularly
+     * usefull in places where MarkDirty is conditionally called and the
+     * conditiont can be expensive and thus can be skipped if the flag is true.
```
Comment (instagibbs) [inline src/wallet/wallet.h]: conditiont?

Comment (promag) [reply src/wallet/wallet.h]: Yes :trollface:

**Comment by achow101** [comment]
ACK c9a314c6dd3bf48f5bc58e606470b39a3c963724

**Comment by promag** [comment]
Fixed typo "conditiont" as reported by @instagibbs.

**Comment by promag** [comment]
Rebased now that base PR #17843 is merged.

**Comment by fjahr** [comment]
Re-ACK fd6d067

**Comment by achow101** [comment]
ACK fd6d0679e7c3a956afc273fa38f64dab3f0ed223

**Review by achow101 — APPROVED** [review]

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/17889/commits/fd6d0679e7c3a956afc273fa38f64dab3f0ed223

though I think better encapsulation of these concepts would be a plus so we can't accidentally mark dirty or set the cache without knowing.

**Review by meshcollider — COMMENTED** [review]
utACK fd6d0679e7c3a956afc273fa38f64dab3f0ed223

**Pull request merged**